File: foreign/go/contracts/binary_request_kind.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package iggcon

// BinaryRequestKind declares how a raw binary request executes on the server.
//
// Classic framing carries [length][code][payload] with no operation field, so
// this SDK encodes both kinds identically and the server decides how a command
// runs. The declaration exists so the API matches every other Iggy SDK and is
// already in place when the Go client gains VSR framing.
type BinaryRequestKind uint8

const (
	// BinaryRequestKindNonReplicated runs on the receiving node only, outside
	// consensus.
	BinaryRequestKindNonReplicated BinaryRequestKind = 1
	// BinaryRequestKindReplicated is replicated through consensus before it
	// takes effect.
	BinaryRequestKindReplicated BinaryRequestKind = 2
)

// String returns the cross-SDK name of the kind.
func (kind BinaryRequestKind) String() string {
	switch kind {
	case BinaryRequestKindNonReplicated:
		return "non_replicated"
	case BinaryRequestKindReplicated:
		return "replicated"
	default:
		return "unknown"
	}
}

// IsValid reports whether the kind is one the API defines. Go has no exhaustive
// enums, so a caller can hand over any byte.
func (kind BinaryRequestKind) IsValid() bool {
	return kind == BinaryRequestKindNonReplicated || kind == BinaryRequestKindReplicated
}

File: core/common/src/types/binary_request_kind.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use strum::{Display, EnumString};

/// How a raw binary request executes on the server.
///
/// A vendor command code is unknown to the SDK, so the caller has to say whether
/// it runs outside consensus or is replicated through it. The declaration is
/// never authoritative: a code the protocol tables already know keeps its own
/// class, and the server independently rejects a declaration it disagrees with.
///
/// Classic framing carries `[length][code][payload]` with no operation field, so
/// the kind is inert there and both variants encode identical bytes. It only
/// selects a wire path under VSR.
///
/// The string form is the cross-SDK spelling the shared BDD scenarios and the
/// PHP binding use.
#[derive(Debug, Copy, Clone, PartialEq, Eq, Display, EnumString)]
#[strum(serialize_all = "snake_case")]
pub enum BinaryRequestKind {
    /// Runs on the receiving node only, outside consensus. Reads, pings, and
    /// vendor commands that own no replicated state.
    NonReplicated,
    /// Replicated through consensus before it takes effect.
    ///
    /// Only the standard replicated commands are supported today. A vendor code
    /// declared `Replicated` yields [`crate::IggyError::FeatureUnavailable`]:
    /// the protocol has no deterministic handler registry, replicated state
    /// ownership, or snapshot contract for one yet.
    Replicated,
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use super::BinaryRequestKind;

    #[test]
    fn given_display_output_when_parsed_should_round_trip() {
        for kind in [
            BinaryRequestKind::NonReplicated,
            BinaryRequestKind::Replicated,
        ] {
            assert_eq!(
                BinaryRequestKind::from_str(&kind.to_string()).unwrap(),
                kind
            );
        }
    }

    #[test]
    fn given_cross_sdk_names_when_parsed_should_match_the_other_bindings() {
        assert_eq!(
            BinaryRequestKind::NonReplicated.to_string(),
            "non_replicated"
        );
        assert_eq!(BinaryRequestKind::Replicated.to_string(), "replicated");
    }

    #[test]
    fn given_unknown_text_when_parsed_should_fail() {
        assert!(BinaryRequestKind::from_str("auto").is_err());
    }
}
